### ISSUE TITLE
Using another dataset

### ISSUE THREAD

**Issue by CatarinaPC** [thunlp/OpenNRE#269]
Hello

I would like to use OpenNRE to train a French model.

1. I already have the data which I obtained by **distant supervision** (DBpedia + Wikipedia) and I'm trying to figure out into what structure I need to get it to be able to use it on OpenNRE.
I looked at the examples in the benchmark folder: all the datasets (except semeval), in each sentence, have an id for each entity. How should I replicate this in my dataset? Generating random unique numbers to represent each entity in my dataset?

2. I know that in the NYT dataset, if an entity has more than one word, its word embedding is obtained by the concatenation of the multiple words. If I want to use French word embeddings do I need to train them to expect this? (did you train your BERT embedding too?)

3. I also noticed that the mentions of the entities in sentences were replaced by their Freebase form (e.g., in a sentence that mentioned Obama, we have instead Barack Obama). Should I do the same for my dataset or leaving the entities as they are mentioned in the text ok (I see there can be possibly be problems when trying to group them in bags)

4. Do the word embeddings I use need to be in any special format?

5. In the examples folder I see examples for training 3 different kinds of models. Do I need to train a specific kind of model because I use distant supervison?

**Comment by gaotianyu1350** [comment]
Hi, 

1. You can generate entity ids however you like, as long as each entity has a unique one.
2. You do not need to train the word embedding to get that. If you have a pre-trained entity embedding, just use it as you have suggested. We have trained a BERT model actually, but since it is too hard to do a bag-level training for BERT (due to CUDA memory limit), we just trained it in a supervised way and the results are pretty good. We plan to update it to the repo recently.
3. I am not sure of this. Maybe the original dataset did so, or maybe when constructing the dataset they were only able to extract entities that had exactly the same names as in the Freebase back then. I think both methods are ok. Replacing the mentions with their unified names has an advantage: the model can implicitly learn a knowledge embedding for them.
4. Just following the example glove file. You need a numpy file storing the word embedding and also a json file specifying the word-id mapping (starting from zero). 
5. You should use `train_bag_pcnn_att.py`.

**Comment by CatarinaPC** [comment]
Thank you so much.

If you could clarify just a few more things I would really appreciate it:

1. Focusing on the (previous) points 1 and 3, if an entity is mentioned in different sentences, in different ways, should I keep the id the same between them? For example, if the head entity in 3 different sentences is, respectively: "Barack", "Obama" and "Barack Obama". Should these head entities have different ids or the same id?

2. 

> You do not need to train the word embedding to get that. If you have a pre-trained entity embedding, just use it as you have suggested.

Can you clarify this part a bit more? I'm a bit confused with the _entity embedding_ part and using as I suggest it. Can I use any pre-trained word embeddings available online, e.g., French fastText embeddings, without having to anything else?

3. From the (previous) points 4. and 5. I think it is not possible to use "dynamic" word embeddings like BERT, and others, in combination with bag+pcnn+att training. Is this true?

Thank you again

**Comment by gaotianyu1350** [comment]
1. Yeah, you should keep the IDs the same.
2. Yes, you can directly use any available word embedding. I mean you can get the entity embedding by averaging embeddings of all the words in the entity mention.
3. Considering the CUDA memory limit, yes, it is almost impossible to use BERT in bag-level training. We manage to train a bag-level BERT by limiting the size of each bag less than 3. However, the result is even lower than BERT without bag-level training.

**Comment by CatarinaPC** [comment]
> I mean you can get the entity embedding by averaging embeddings of all the words in the entity mention.

But that would need to be already in your code, right? Because when creating the model it expects a word2id kind of file. So, the averaging would have to be after.

As I said before, I'm using DBpedia as a KB and Wikipedia as my textual source. Bags make sense in distant supervision. I have been reading and saw some people creating a dataset also using Wikipedia where they only consider the first sentence (bags stop making sense). I'm thinking of, in addition to bag training, also doing this and maybe experimenting with the other types of models. Do you have any feedback?

**Comment by gaotianyu1350** [comment]
Yeah it totally depends on you. You can get the entity embedding by averaging word embeddings. Note that in our current implementation, we do not use this method: we use the original text and we do not merge each entity as a single token when training. It has its own benefits, especially when using BERT. However, if you would like to get the entity embedding by averaging, you can modify the code.


As for the second question, you can definitely try it. From my side, simply using BERT without any bag-level training achieves best results and you should probably try that.
On 08/17/2020 22:05, Catarina Conceição wrote:

I mean you can get the entity embedding by averaging embeddings of all the words in the entity mention.

But that would need to be already in your code, right? Because when creating the model it expects a word2id kind of file. So, the averaging would have to be after.

As I said before, I'm using DBpedia as a KB and Wikipedia as my textual source. Bags make sense in distant supervision. I have been reading and saw some people creating a dataset also using Wikipedia where they only consider the first sentence (bags stop making sense). I'm thinking of, in addition to bag training, also doing this and maybe experimenting with the other types of models. Do you have any feedback?

**Comment by CatarinaPC** [comment]
I'm sorry again.

Still in relation to the ID's of the entities. Looking at the examples in the benchmark folder, for each entity in each sentence we have, for example: 

`"h": {"name": "fort lauderdale-hollywood international airport", "id": "Q635361", "pos": [9, 15]}`

As we have talked in this issue, I should keep the ids the same between different mentions of the same entity:

> For example, if the head entity in 3 different sentences is, respectively: "Barack", "Obama" and "Barack Obama". Should these head entities have different ids or the same id?

But, when it comes to 'name', is it supposed to correspond to how the entity is mentioned in the text (e.g., "Barack", "Obama" or "Barack Obama") or should it be almost like the id, where the name is the same between different sentences even if its mention in the text is different (e.g., "Barack Obama") ?

**Comment by gaotianyu1350** [comment]
For the `name` key in `h` and `t`, since we do not use this in our code (we use `id` keys to distinguish entities), you can assign it however you like. Conventionally we would assign the `name` key as the entity mentions.

**Comment by CatarinaPC** [comment]
As for the sentence, if can be in a key 'token' or a key 'text'.

I want to provide my sentences already tokenized, however, I have the positions of the entities in the sentence character-wise .

If I use 'text' (tokenize my sentence, and join the strings using whitespace) does the system perform whitespace tokenization or another type of tokenization?

If its the later i have to use 'token' and find a way to convert the character-wise positions to list positions.

Thank you

**Comment by gaotianyu1350** [comment]
For PCNN-ATT, the system uses whitespace tokenization. For BERT it uses the BERT tokenizer.

**Comment by CatarinaPC** [comment]
Thank you.

But if I use 'token' is what you described the same? Will it, for example, when using BERT, "re-tokenize" the input?

**Comment by gaotianyu1350** [comment]
1. If you use BERT and `token`, OpenNRE will use BERT tokenizer to further tokenized each token. 

2. For now we did not adopt OpenNRE to other transformer models. However, since they have similar interfaces in huggingface's `transformers`, you can easily add them yourself. To add a model, you can refer to `opennre/encoder`. For each encoder its tokenization method is in `tokenize` function.

3. I think you need to create your own `word2id.json` file. The `id` is the index of the word in the npy file.

4. You should keep the `bag_size` as 0. When assigned a positive number, `bag_size` means only using `bag_size` number of instances for each bag.

**Comment by CatarinaPC** [comment]
Thank you again for answering all of my questions.

1. I wanted to ask if it is possible to evaluate the model after training, on a "separate script". Because, in the example file I noticed the variable `framework` (object of `BagRE`) is necessary:

```
(...)

# Test the model
framework.load_state_dict(torch.load(ckpt)['state_dict'])
result = framework.eval_model(framework.test_loader)

# Print the result
logging.info('Test set results:')
logging.info('AUC: {}'.format(result['auc']))
logging.info('Micro F1: {}'.format(result['micro_f1']))
```

If it it possible how could it be achieved?

2. Also, just to confirm, in distant supervision the typical metrics one would present when evaluating a model are what? Precision-recall curve and precision P@N ? Or, typical multi-class classification metrics like precision, recall, f1-score, etc?

3. Have you found that hyperparameterization selection is worth experimenting with, i.e., there are significant increases in scores, or using the default parameterization is enough?

4. The default metric for picking up best checkpoint is "auc". Do you find it better or equivalent to using micro_f1?

5. Going back to a previous answer of yours:

> As for the second question, you can definitely try it. From my side, simply using BERT without any bag-level training achieves best results and you should probably try that.

When you say _"simply using BERT without any bag-level training"_ do you mean that, for example you trained and evaluated the NYT dataset **as is** (more than one sentence per triple (ent1,ent2, rel)) with both BERT sentence-level and bag-training, and BERT sentence-level was better?

6. Roughly, how long does it take to train a model, like the pre-trained ones that are available?

**Comment by gaotianyu1350** [comment]
1. You can use `--only_test` arg in the example code to only test the model.

2. We usually use `auc` (area under the curve) to measure distant supervision models.

3. I think you should try different hyper-parameters for your dataset, as it may significantly affect the results. What I provided here is only (maybe) the best for nyt10 dataset.

4. I think `micro_f1` is similar to `auc`. However, you can try it out to see if there is a big difference.

5. Yes. Sentence-level training on BERT was better.

6. For pcnn-att on NYT10, it took roughly two or three hours on a single 2080ti gpu. For BERT on tacred / wiki80, it roughly takes two hours on a single 2080ti.
